### PR TITLE
Assert messages to flush to the log file

### LINKED ISSUE #913: [FEATURE] How to ensure that `assert` messages are logged?
### Is your feature request related to a problem? Please describe.
Currently, all messages from the `assert` conditions flush to the console, which is fine but it does not present a unified logging interface.

### Describe the solution you'd like
It would be great if those messages could get flushed to the log file (perhaps in addition to the console).

### Describe alternatives you've considered
N.A.

### Additional context
Perhaps this could be a solution: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertLogs

### PR DESCRIPTION
**Pull request by benmalef**
Fixes #913 

## Proposed Changes

-  Updated gandlf_logging to flush the assert logs into the file using the  [sys.excepthook](https://docs.python.org/3/library/sys.html#sys.excepthook) 

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

### CODE REVIEW

**Comment by sarthakpati** [comment]
This is good to go, right, @benmalef?

**Comment by sarthakpati** [comment]
@VukW do you have any feedback?

**Comment by VukW** [comment]
Hey folks, from a first glance it seems to be a legit way to log an exception, but to tell truth I'm confused (I've never used `excepthook` previously tho). let me check it a bit more deeper, cos I'm not 100% sure it should be used in this way

**Comment by sarthakpati** [comment]
> Hey folks, from a first glance it seems to be a legit way to log an exception, but to tell truth I'm confused (I've never used `excepthook` previously tho). let me check it a bit more deeper, cos I'm not 100% sure it should be used in this way

Any update, @VukW?

File: GANDLF/utils/gandlf_logging.py [hunk 1]
```diff
@@ -26,6 +28,16 @@ def _configure_logging_with_logfile(log_file, config_path):
         logging.config.dictConfig(config_dict)
 
 
+def gandlf_excepthook(exctype, value, tb):
+    if issubclass(exctype, AssertionError):
```
Comment (VukW) [inline GANDLF/utils/gandlf_logging.py:32]: Despite of the initial issue is only about Assertions, let's log absolutely any uncaught exceptions here?

File: GANDLF/utils/gandlf_logging.py [hunk 2]
```diff
@@ -26,6 +28,16 @@ def _configure_logging_with_logfile(log_file, config_path):
...
+        logging.exception(
+            f"{exctype.__name__}: {value}\n{''.join(traceback.format_exception(exctype, value, tb))}",
+            exc_info=False,
+        )
```
Comment (VukW) [inline GANDLF/utils/gandlf_logging.py:36]:
```suggestion
        logging.exception("Uncaught exception", exc_info=(exctype, value, tb))
```
`Logging` can pretty-print exceptions on its own. Ow do we need a special formatting?

**Comment by VukW** [comment]
@benmalef does PR work for you? For me `logging.exception` is not catch to the debug file...

**Comment by benmalef** [comment]
> @benmalef does PR work for you? For me `logging.exception` is not catch to the debug file...

@VukW How do you test it? 
 It was fine when I tested it on my PC, but I'll check it again.
 Is logging.exception generally not caught, or is this only an issue in the excepthook?
Then, maybe this isn't the right approach.

**Comment by benmalef** [comment]
## Important Notes 

I have tried to use the [sys.excepthook](https://docs.python.org/3/library/sys.html#sys.excepthook) to change the behavior of the assert statement. When a AssertionError is raised, it should be flushed in the console and saved in the log file.  In  my PC , it works well, at least with simple situations, but @VukW has mentioned that it is not working properly.  So maybe is not the right approach. 

### Testing
I have tried to test it,  but the pytest has its implementation of managing the AssertionError. 
One solution maybe is  [this](https://stackoverflow.com/questions/29036354/custom-sys-excepthook-doesnt-work-with-pytest).  I tried it but I didn't figure it out.
Another way that I have tried is to change the assert statement behavior in pytest. I followed some of these steps [ref](https://stackoverflow.com/questions/58321991/is-it-possible-to-change-pytests-assert-statement-behaviour-in-python) , but I got the same result. 

To cancel the default behavior it can be done using [this](https://docs.pytest.org/en/stable/how-to/assert.html#disabling-assert-rewriting)

**Comment by sarthakpati** [comment]
@benmalef thank you for looking into this. 

@VukW: it would be great if you could provide some feedback. Until then, I would recommend we keep this PR as draft.

**Comment by sarthakpati** [comment]
> everything work exactly as expected now for me

I think with this, we can proceed with the merge. I'll wait for the tests to finish.

**Comment by VukW** [comment]
@sarthakpati sure, i don't have any other comments except already mentioned ones

File: GANDLF/utils/gandlf_logging.py (hunk 1 above)
Comment (sarthakpati) [reply GANDLF/utils/gandlf_logging.py:32]: Should the check be the following, instead?

```python
if issubclass(exctype, Exception)
```

File: GANDLF/utils/gandlf_logging.py [hunk 3]
```diff
@@ -4,6 +4,8 @@
 from importlib import resources
 import tempfile
 from GANDLF.utils import get_unique_timestamp
+import sys
+import traceback
```
Comment (sarthakpati) [inline GANDLF/utils/gandlf_logging.py:8]: Fix codacy error.

```suggestion
```

File: GANDLF/utils/gandlf_logging.py (hunk 1 above)
Comment (benmalef) [reply GANDLF/utils/gandlf_logging.py:31]: Yes, If we want to log all the uncaught exceptions.

Comment (sarthakpati) [reply GANDLF/utils/gandlf_logging.py:31]: Great, then let's make that change and get ready to merge! 😄

**Review by sarthakpati — APPROVED** [review]

**Pull request merged**